src/components/stack-chart/Canvas.js

```diff
@@ -382,72 +383,21 @@ class StackChartCanvas extends React.PureComponent<Props> {
-    const {
-      thread,
-      stackTimingByDepth,
-      categories,
-      callNodeInfo: { callNodeTable },
-    } = this.props;
-    const stackTiming = stackTimingByDepth[depth];
+    const { thread, stackTimingByDepth, categories, callNodeInfo } = this.props;
 
+    const stackTiming = stackTimingByDepth[depth];
+    const callNodeIndex = stackTiming.callNode[stackTimingIndex];
     const duration =
       stackTiming.end[stackTimingIndex] - stackTiming.start[stackTimingIndex];
 
-    const callNodeIndex = stackTiming.callNode[stackTimingIndex];
-    const categoryIndex = callNodeTable.category[callNodeIndex];
-    const category = categories[categoryIndex];
-    const funcIndex = callNodeTable.func[callNodeIndex];
-    const funcStringIndex = thread.funcTable.name[funcIndex];
-    const funcName = thread.stringTable.getString(funcStringIndex);
-
-    let resourceOrFileName = null;
-    // Only JavaScript functions have a filename.
-    const fileNameIndex = thread.funcTable.fileName[funcIndex];
-    if (fileNameIndex !== null) {
-      // Because of our use of Grid Layout, all our elements need to be direct
-      // children of the grid parent. That's why we use arrays here, to add
-      // the elements as direct children.
-      resourceOrFileName = [
-        <div className="tooltipLabel" key="file">
-          File:
-        </div>,
-        thread.stringTable.getString(fileNameIndex),
-      ];
-    } else {
-      const resourceIndex = thread.funcTable.resource[funcIndex];
-      if (resourceIndex !== -1) {
-        const resourceNameIndex = thread.resourceTable.name[resourceIndex];
-        if (resourceNameIndex !== -1) {
-          // Because of our use of Grid Layout, all our elements need to be direct
-          // children of the grid parent. That's why we use arrays here, to add
-          // the elements as direct children.
-          resourceOrFileName = [
-            <div className="tooltipLabel" key="resource">
-              Resource:
-            </div>,
-            thread.stringTable.getString(resourceNameIndex),
-          ];
-        }
-      }
-    }
-
     return (
-      <div className="stackChartCanvasTooltip">
-        <div className="tooltipOneLine tooltipHeader">
-          <div className="tooltipTiming">{formatNumber(duration)}ms</div>
-          <div className="tooltipTitle">{funcName}</div>
-        </div>
-        <div className="tooltipDetails">
-          <div className="tooltipLabel">Category:</div>
-          <div>
-            <span
-              className={`category-swatch category-color-${category.color}`}
-            />
-            {category.name}
-          </div>
-        </div>
-        <div className="tooltipDetails">{resourceOrFileName}</div>
-      </div>
+      <TooltipCallNode
+        thread={thread}
+        callNodeIndex={callNodeIndex}
+        callNodeInfo={callNodeInfo}
+        categories={categories}
+        durationText={`${formatNumber(duration)}ms`}
```

Comment:
I think we can use `formatMilliseconds` here instead.

Reply:
Whoops! Good call.